File: ai-service/app/utils/metrics.py
"""
metrics.py
==========
In-process latency tracking for the AI Service — Sprint 10, Part 4
(Monitoring). Mirrors server/src/middlewares/metrics.js on the Node side:
an in-memory rolling window per pipeline category, not a Prometheus/StatsD
integration — appropriate for a single-instance MVP, documented as tech
debt for a multi-instance deployment.
"""

import time
from collections import deque
from statistics import mean

WINDOW_SIZE = 200

_samples: dict[str, deque] = {}
_totals = {"requests": 0, "errors": 0}

CATEGORY_PREFIXES = {
    "/predict": "prediction",
    "/explain": "shap",
    "/employee-intelligence": "employee_intelligence",
    "/knowledge": "knowledge",
    "/rag": "knowledge",
    "/decision": "decision",
}


def _categorize(path: str) -> str | None:
    for prefix, category in CATEGORY_PREFIXES.items():
        if path.startswith(prefix):
            return category
    return None


def _percentile(values: list[float], p: float) -> float | None:
    if not values:
        return None
    sorted_values = sorted(values)
    index = min(len(sorted_values) - 1, int((p / 100) * len(sorted_values)))
    return round(sorted_values[index], 1)


async def metrics_middleware(request, call_next):
    start = time.perf_counter()
    response = await call_next(request)
    duration_ms = (time.perf_counter() - start) * 1000

    _totals["requests"] += 1
    if response.status_code >= 500:
        _totals["errors"] += 1

    category = _categorize(request.url.path)
    if category:
        if category not in _samples:
            _samples[category] = deque(maxlen=WINDOW_SIZE)
        _samples[category].append(duration_ms)

    return response


def get_metrics_snapshot() -> dict:
    categories = {}
    for name, values in _samples.items():
        values_list = list(values)
        categories[name] = {
            "sampleCount": len(values_list),
            "avgMs": round(mean(values_list), 1) if values_list else None,
            "p95Ms": _percentile(values_list, 95),
            "maxMs": round(max(values_list), 1) if values_list else None,
        }
    total_requests = _totals["requests"]
    total_errors = _totals["errors"]
    return {
        "totalRequests": total_requests,
        "totalErrors": total_errors,
        "errorRate": round((total_errors / total_requests) * 100, 2) if total_requests else 0,
        "categories": categories,
    }

File: server/src/middlewares/metrics.js
/**
 * @file metrics.js
 * @description In-process latency tracking — Sprint 10, Part 4 (Monitoring).
 *
 * Deliberately in-memory rather than Prometheus/StatsD: this is a
 * single-instance MVP, so a rolling sample window per route category is
 * enough to answer "is prediction/decision/knowledge/dashboard latency
 * healthy right now" from the /health/deep endpoint without adding a new
 * infrastructure dependency. A multi-instance deployment would need to
 * replace this with a real metrics backend — documented as tech debt.
 */

const WINDOW_SIZE = 200;

const CATEGORY_MATCHERS = [
  { name: 'prediction', test: (p) => p.startsWith('/api/v1/ai/predict') },
  { name: 'decision', test: (p) => p.startsWith('/api/v1/decisions') },
  { name: 'knowledge', test: (p) => p.startsWith('/api/v1/knowledge') },
  { name: 'shap', test: (p) => p.startsWith('/api/v1/explain') },
  { name: 'dashboard', test: (p) => p.startsWith('/api/v1/analytics') || p.startsWith('/api/v1/executive/dashboard') || p.startsWith('/api/v1/workflow/dashboard') },
];

const samples = new Map(); // category -> number[] (ms)
let totalRequests = 0;
let totalErrors = 0;

function categorize(path) {
  const match = CATEGORY_MATCHERS.find((c) => c.test(path));
  return match?.name || null;
}

function record(category, durationMs) {
  if (!samples.has(category)) samples.set(category, []);
  const arr = samples.get(category);
  arr.push(durationMs);
  if (arr.length > WINDOW_SIZE) arr.shift();
}

function percentile(arr, p) {
  if (arr.length === 0) return null;
  const sorted = [...arr].sort((a, b) => a - b);
  const index = Math.min(sorted.length - 1, Math.floor((p / 100) * sorted.length));
  return Math.round(sorted[index]);
}

export function metricsMiddleware(request, response, next) {
  const start = process.hrtime.bigint();
  response.on('finish', () => {
    const durationMs = Number(process.hrtime.bigint() - start) / 1e6;
    totalRequests += 1;
    if (response.statusCode >= 500) totalErrors += 1;
    const category = categorize(request.path);
    if (category) record(category, durationMs);
  });
  next();
}

export function getMetricsSnapshot() {
  const categories = {};
  for (const [name, arr] of samples.entries()) {
    categories[name] = {
      sampleCount: arr.length,
      avgMs: arr.length ? Math.round(arr.reduce((a, b) => a + b, 0) / arr.length) : null,
      p95Ms: percentile(arr, 95),
      maxMs: arr.length ? Math.round(Math.max(...arr)) : null,
    };
  }
  return {
    totalRequests,
    totalErrors,
    errorRate: totalRequests > 0 ? Math.round((totalErrors / totalRequests) * 10000) / 100 : 0,
    categories,
  };
}
